addons/website_sale_delivery/controllers/main.py

```diff
@@ -32,3 +33,19 @@ def order_2_return_dict(self, order):
             if line.is_delivery:
                 ret['transaction']['shipping'] = line.price_unit
         return ret
+
+    def _get_shop_payment_values(self, order, **kwargs):
+        values = super(WebsiteSaleDelivery, self)._get_shop_payment_values(order, **kwargs)
+        if not order._get_delivery_methods():
+            values['errors'].append(
+                (_('Sorry, we are unable to ship your order'),
+                 _('No shipping method is available for your current order and shipping address. '
+                   'Please contact us for more information.')))
+
+        has_stockable_products = any(order.order_line.filtered(lambda line: line.product_id.type in ['consu', 'product']))
```

Comment:
Avoiding `filtered` and performing check in `any` is shorter and could be more efficient as it will stop as soon as it fines a matching line rather than filter all the lines then check:
```python
any(line.product_id.type in ['consu', 'product'] for line in order.order_line)
```